### PR TITLE
Use C++20 function std::popcount

### PR DESCRIPTION
**Pull request by a-noni-mousse**
## High Level Overview of Change

Remove the custom code and special compiler functions and use C++20 function [`std::popcount`](https://en.cppreference.com/w/cpp/numeric/popcount) and make `inline` for optimization of the code. The changed code is good when compiled and  produces same or even better assembly code which you can see on the [compiler explorer](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOykrgAyeAyYAHI%2BAEaYxCAAnKQADqgKhE4MHt6%2BASlpGQKh4VEssfFJdpgOmUIETMQE2T5%2BgVU1AnUNBMWRMXGJtvWNzbltwz1hfWUDCQCUtqhexMjsHOYAzGHI3lgA1CYbbtGEh9gmGgCCm9u7mAdHyAoE%2BIJnF9dXN1QMWFR7AH0ALJCNwAgBq2AASh8wgQ9qlkIIuAA2AGedAQZ7oEAgLxw1EA%2BFMOYfEz%2BKxXPbUvbETAEZYMQEA6L42iOBgA1LJNBeQQQEmHSnXfwAETJZg2rjwVDJVzhCNQPORaPCAHcsS9cfiVUS9oKvhSPjTafTGXtsbjubz%2BYKNsLyaKOAtaJwAKy8PwcLSkVCcNzWawWpYre6bHikAiaZ0LADWIEkAA4AHRmDRcfwbMyJ9NmSQJFH6TiSXgsCQaDSkL0%2Bv0cXgKECVqPe52kOCwJBoFjJOhxciULs9%2BjxYBcMxmUhYABueFW4LwmDVAHlkoxOBGaOy4g2INFo6QTsxiABPde8Q8NY9L6LaarNiNdtiCJcMWinluTzAsQzAcQf/B0jUU6YA2H6YKo1ReAQawRnCmCuh%2BtB4NExCXh4WD7gQxB4GW3CtlQBjAAo86Liua54TIggiGI7BSJR8hKGo%2B66Fw%2Bg/iggaWPoyENpACxKo4AigQAtNihyiqYljWGmezCUuZj1vBd6ZC4vyjH4rEhFMpTlHoqTpIJWSeC0ekFIZvQ6QMrHtIZXQjMZuTWUpHQMHZkwlP08TWRM6l6M83QWZ5EgLAoIarMFRYcB6Vb7rWeyqImKLCSikh7MAyDIHsY6pnsEC4IQJAHJKXBzLwzZaAsECdqg3a9mQFAQIOdUoAYRhSBWn4znOC7LquXobnQ0HEDue4fheJ5ngeYSXtet4OJNj6MAQL5vvuWDfkYf4%2BgBynAaBPrgZB0GTXBCE%2BkhKFoRgaw%2BlhOFngsBFMERJG9eRG6yNR4h0fwgiKCo6gfroE6tcYnE2BdvEQPxySGSJYkbBJ4MyXJGyKfYhmqe4DkaUEvyBTMXn5AZmS%2Bax%2BmFAwBO6U5GO1D5ON%2Bc5tkTNTVlDN0ZMc40bNeSFYW0S67qerFnDxYlyWpTsP5ZZIyYaPLuX5UQxBFRsJVldGlXVbVw79o1NVDgMU7IMkyQAlOXAJACZjW%2BCQiElwXAAqoKWdbOmCvWR/W8JuQ0jfu43vhGQezXeC01U%2By2vu%2B21fj%2BW28DtQEgfuh3IFBMG8Kd%2B4XahJ7oTdZXYbhPCPYRxE9d7k2/cIojfdItf/UxQMgBsbFGBxUlcZD8Aw3DnAAPSWpJViWDJg/yejynOBAri%2BdmePoLzbfE5TC%2BJmv5naUF7c2fTnOM4v%2B%2BdKzO%2BE6v/n2TkfjH2fHkXxs/PLOFJWRdF1a8HFrupSwCgm3sS2CRUwJD2PbVEyYnZK3wCrNWGtIxazjCAN0XBkz%2BASImMwKJ/BW0TP4N0GhJCsQQiWGKH5az1hABOcqMZIoKTITWTgmsWxzAWMBYaKlJBAA%3D%3D)

### Type of Change

- [X] Refactor (non-breaking change that only restructures code)

### CODE REVIEW

**Comment by intelliot** [comment]
CI has reported failures; please [review them](https://github.com/XRPLF/rippled/actions/runs/4346233896/jobs/7592024677).

**Comment by HowardHinnant** [comment]
This change is required to compile on macOS: https://github.com/HowardHinnant/rippled/commit/ab5e386a39e9f77c91051262e06ca80127d9552a

Explanation:  The expression `isBranch_ | (1 << m)` promotes to `int` which is then sent to `std::popcount`.  But `popcount` has a [constraint that its argument be an unsigned integral type](http://eel.is/c++draft/bit#count-10).  Adding the `u` suffix makes the expression `unsigned`.

**Comment by intelliot** [comment]
@a-noni-mousse please review the comments above.

**Comment by intelliot** [comment]
@a-noni-mousse for future reference - there's no need to force push, as this PR will be squashed when it is merged. 

Keeping historical commits may help reviewers to more easily see how the PR has changed over time.

**Review by nbougalis — APPROVED** [review]
Looks fine.

**Comment by intelliot** [comment]
@a-noni-mousse - going forward, please sign your commits, as mentioned in [CONTRIBUTING](https://github.com/XRPLF/rippled/blob/develop/CONTRIBUTING.md). (Past commits are fine to leave as-is)

File: src/ripple/shamap/impl/TaggedPointer.h [hunk 1]
```diff
@@ -217,6 +217,31 @@ class TaggedPointer
     getChildIndex(std::uint16_t isBranch, int i) const;
 };
 
+[[nodiscard]] inline int
+popcnt16(std::uint16_t a)
+{
+#if __cpp_lib_bitops
+    return std::popcount(a);
+#else if defined(__clang__) || defined(__GNUC__)
```
Comment (HowardHinnant) [inline src/ripple/shamap/impl/TaggedPointer.h:225]:
```
/Users/howardhinnant/Development/rippled/src/ripple/shamap/impl/TaggedPointer.h:225:7: warning: extra tokens at end of #else directive [-Wextra-tokens]
#else if defined(__clang__) || defined(__GNUC__)
      ^
      //
/Users/howardhinnant/Development/rippled/src/ripple/shamap/impl/TaggedPointer.h:227:2: error: #else after #else
#else
 ^
```

File: src/ripple/shamap/impl/TaggedPointer.h [hunk 2]
```diff
@@ -217,6 +217,31 @@ class TaggedPointer
     getChildIndex(std::uint16_t isBranch, int i) const;
 };
 
+[[nodiscard]] inline int
```
Comment (HowardHinnant) [inline src/ripple/shamap/impl/TaggedPointer.h:220]: Missing `#include <bit>`

Missing `#include <array>`

**Review by HowardHinnant — CHANGES_REQUESTED** [review]

**Comment by intelliot** [comment]
There's a branch protection rule that requires commits to be signed in order to be merged. It can be bypassed, but it's preferable to sign commits to provide some assurance about their integrity.

**Review by HowardHinnant — APPROVED** [review]

**Review by mtrippled — APPROVED** [review]
LGTM :+1:

**Comment by intelliot** [comment]
Suggested commit message for squashed commit:
```
refactor: use C++20 function std::popcount (#4389)

- Replace custom popcnt16 implementation with std::popcount from C++20
- Maintain compatibility with older compilers and MacOS by providing a
  conditional compilation fallback to __builtin_popcount and a lookup
  table method
- Move and inline related functions within SHAMapInnerNode for
  performance and readability
```

@manojsdoshi - be sure to squash this PR when merging, as requested above.

**Pull request closed without merging**